src/value.jl

```diff
@@ -1,145 +1,181 @@
-function CategoricalValue(level::Integer, pool::CategoricalPool{T, R}) where {T, R}
-    return CategoricalValue(convert(R, level), pool)
+# "categorical value" trait
+abstract type CatValueSupport end
+struct IsCatValue <: CatValueSupport end
+struct NotCatValue <: CatValueSupport end
+
+# by default types do not have "categorical value" trait
+iscatvalue(::Type) = NotCatValue
+iscatvalue(::Type{Union{}}) = NotCatValue # otherwise it dispatches to Type{<:...}
+iscatvalue(x::Any) = iscatvalue(typeof(x))
+
+# union of all types that have "categorical value" trait
+const CatValue{R} = Union{CategoricalValue{T, R} where T,
+                          CategoricalString{R}}
+
+# "categorical value" trait implementation for CategoricalValue and CategoricalString
+iscatvalue(::Type{<:CatValue}) = IsCatValue
+valtype(::Type) = Union{} # no value type if not a proper categorical value type
+valtype(::Type{<:CategoricalValue{T}}) where {T} = T
+valtype(::Type{<:CategoricalString}) = String
+
+# integer type of category reference codes for given type
+reftype(::Type) = DefaultRefType
+# TODO reftype(::Type{T}) where {T <: Integer} = T ?
+reftype(::Type{<:CatValue{R}}) where {R} = R
+
+pool(x::CatValue) = x.pool
+level(x::CatValue) = x.level
+
+# extract the type of the original value from array eltype `T`
+unwrap_catvalue_type(::Type{T}) where {T} =
+    unwrap_catvalue_type(iscatvalue(T), T)
```

Comment:
Using `iscatvalue(T)` here is really overkill when you could just dispatch on `<:CatValue`.